scripts/deploy.py

```diff
@@ -0,0 +1,175 @@
+import re
+import inspect
+from typing import Never
+from pathlib import Path
+
+import click
+
+
+def err(text: str) -> Never:
+    sys.stderr.write(text)
+    sys.exit(1)
+
+
+def run(*args: str) -> None:
+    subprocess.run(args, check=True)
+
+
+def run_capture(*args: str) -> str:
+    proc = subprocess.run(args, check=True, stdout=subprocess.PIPE, stderr=None)
+    return proc.stdout.decode("utf-8").strip()
+
+
+def log(msg: str) -> None:
+    print(f"> {msg}")
+
+
+@click.command()
+@click.option(
+    "--rsync-target",
+    help="Where bootstrap zipballs and source tarball will be copied to. If given, this is passed directly to rsync so it can be a local folder, ssh path, etc. For production builds this should be a webroot directory that will be served at bootstrap-url",
+)
+@click.option(
+    "--bootstrap-url",
+    help="URL where bootstrap zip files are available. Defaults to folder part of public-url if not given.",
+)
+@click.option(
+    "--arches",
+    default="aarch64,x86_64",
+    help="Which architectures to build for, comma-separated.",
+)
+@click.argument("public-url")
+def go(
+    public_url: str,
+    rsync_target: str | None,
+    bootstrap_url: str | None,
+    arches: str,
+) -> None:
+    """
+    Builds bootstrap zip balls and source code tar ball (for usage as a channel or flake). If rsync_target is specified, uploads it to the directory specified in rsync_target. The contents of this directory should be reachable by the android device with public_url.
+
+    Examples:
+
+    \b
+    $ nix run .#deploy -- \\
+        'https://example.com/bootstrap/source.tar.gz' \\
+        --rsync-target 'user@host:/path/to/bootstrap'
+
+    \b
+    $ nix run .#deploy -- \\
+        'github:USER/nix-on-droid/BRANCH' \\
+        --rsync-target 'user@host:/path/to/bootstrap' \\
+        --bootstrap-url 'https://example.com/bootstrap/'
+
+    \b
+    $ nix run .#deploy -- \\
+        'file:///data/local/tmp/n-o-d/archive.tar.gz'
+
+    ^ useful for testing. Note this is a path on the android device running the APK, not on the build machine
+    """
+    repo_dir = run_capture(GIT, "rev-parse", "--show-toplevel")
+    os.chdir(repo_dir)
+    source_file = "source.tar.gz"
+    if (m := re.search("^github:(.*)/(.*)/(.*)", public_url)) is not None:
+        channel_url = f"https://github.com/{m[1]}/{m[2]}/archive/{m[3]}.tar.gz"
+        if bootstrap_url is None:
+            err("--botstrap-url must be provided for github URLs")
+    elif re.search("^(https?|file)://", public_url):
+        channel_url = public_url
+    else:
+        err(f"unsupported url {public_url}")
+
+    # for CI and local testing
+    if (m := re.search("^file:///(.*)/archive.tar.gz$", public_url)) is not None:
+        flake_url = f"/{m[1]}/unpacked"
+    else:
+        flake_url = public_url
+    base_url = re.sub("/[^/]*$", "", public_url)
+    if bootstrap_url is None:
+        bootstrap_url = base_url
+
+    log(f"channel_url = {channel_url}")
```

Comment:
sorry, were these four calculated and then never used?

Reply:
yes, oops. will fix when i get a chance

Reply:
fixed